storybookjs/storybook
# Angular: Improve decorators

Issue: #12749 #10272

I think this solves this issue:
https://github.com/storybookjs/storybook/issues/12749

And gives a solution for this one
https://github.com/storybookjs/storybook/issues/10272 

Remove this note in decorators documentation :
> // Note: To see the global decorator applied to your stories you'll need to include the template key with the component's selector with each story you have.


## What I did

- Transforms the component (if present) into a template so that it can be included in other templates. 
- Add of 2 decorators function for ease of use


<details><summary>Exemple before/after</summary>
Before: 

```js
export default {
  title: 'Example/Button',
  component: ButtonComponent,
  decorators: [
    (storyFunc) => {
      const story = storyFunc();

      return {
        ...story,
        template: `<div style="margin: 3em">${story.template}</div>`,
      };
    },
  ],
} as Meta;

// Note:  To see the decorator applied to your component's stories you'll need to include the template key with the component's selector.

const Template: Story<ButtonComponent> = (args: ButtonComponent) => ({
  component: ButtonComponent,
  moduleMetadata: {
    declarations: [ButtonComponent],
  },
  props: args,
  template: '<app-button label="Submit"></app-button>',
});
```

After :
```js
export default {
  title: 'Example/Button',
  component: ButtonComponent,
  decorators: [
    moduleMetadata({
      declarations: [ParentComponent],
    }),
    // With template
    componentWrapperDecorator((stroy) => `<div style="margin: 3em">${story}</div>`),
    // With component which contains ng-content
    componentWrapperDecorator(ParentComponent),
  ],
} as Meta;

const Template: Story<ButtonComponent> = (args: ButtonComponent) => ({
  props: args,
  template: '<app-button label="Submit"></app-button>',
});

// !! this was not possible before 
const Component: Story<ButtonComponent> = (args: ButtonComponent) => ({
  props: args,
});
```

</details>


<details><summary>Exemple with ng-content</summary>

```js 
// ParentComponent
@Component({
  selector: 'parent-component',
  template: `<div style="margin: 3em; border:solid;"><ng-content></ng-content></div>`,
})
export default class ParentComponent {}

// Story 

export default {
  title: 'Example/Button',
  /**
To get the docs of parent component.
But needs to add a template in the Story otherwise storybook will take this component here by default 
We can also put child component according to story goal
*/
  component: ParentComponent, 
  decorators: [
    componentWrapperDecorator(ParentComponent),
  ],
} as Meta;

const EmptyContent = (args) => ({
  props: args,
  template: '',
});

const WithContent = (args) => ({
  props: args,
  template: 'The content',
});


```
</details>


## How to test

- Is this testable with Jest or Chromatic screenshots? na
- Does this need a new example in the kitchen sink apps? na
- Does this need an update to the documentation? yes

-> Go on each story to check them 
-> Thanks to the storyshots tests it is clear that the behaviour does not change 🤞 

## TODO

- [x] (TODO in another PR) ~~Improve moduleMetadata decorator type~~
- [x] (TODO in another PR) ~~move preview/angular/decorators.ts in /preview/decorators.ts~~
- [x] add props (with args) in decorator function
